daffodil-io/src/main/scala/org/apache/daffodil/io/InputSourceDataInputStream.scala

```diff
@@ -0,0 +1,840 @@
+
+  def reset(mark: DataInputStream.Mark): Unit = {
+    val current = releaseUntilMark(mark)
+    Assert.invariant(current eq mark)
+    cst.assignFrom(current)
+    setBitPos0b(cst.bitPos0b)
+    inputSource.releasePosition(current.bytePos0b)
+    markPool.returnToPool(current)
+  }
+
+  def discard(mark: DataInputStream.Mark): Unit = {
+    val current = releaseUntilMark(mark)
+    Assert.invariant(current eq mark)
+    inputSource.releasePosition(current.bytePos0b)
+    markPool.returnToPool(current)
+  }
+
+  override def markPos: MarkPos = bitPos0b
+  override def resetPos(m: MarkPos) {
+    setBitPos0b(m)
+  }
+
+  def validateFinalStreamState {
+    // threadCheck()
+    markPool.finalCheck
+  }
+
+  final def getString(nChars: Long, finfo: FormatInfo): Maybe[String] = {
+    val startingBitPos = bitPos0b
+    val aligned = align(finfo.encodingMandatoryAlignmentInBits, finfo)
+    if (!aligned) {
+      Maybe.Nope
+    } else {
+      withLocalCharBuffer { lcb =>
+        val cb = lcb.getBuf(nChars)
+        val numDecoded = finfo.decoder.decode(this, finfo, cb)
+        if (numDecoded == nChars) {
+          Maybe(cb.flip.toString)
+        } else {
+          setBitPos0b(startingBitPos)
+          Maybe.Nope
+        }
+      }
+    }
+  }
+
+  final def getSomeString(nChars: Long, finfo: FormatInfo): Maybe[String] = {
+    val startingBitPos = bitPos0b
+    val aligned = align(finfo.encodingMandatoryAlignmentInBits, finfo)
+    if (!aligned) {
+      Maybe.Nope
+    } else {
+      withLocalCharBuffer { lcb =>
+        val cb = lcb.getBuf(nChars)
+        val numDecoded = finfo.decoder.decode(this, finfo, cb)
+        if (numDecoded > 0) {
+          Maybe(cb.flip.toString)
+        } else {
+          setBitPos0b(startingBitPos)
+          Maybe.Nope
+        }
+      }
+    }
+  }
+
+  lazy val skipCharBuffer = CharBuffer.allocate(32)
+
+  def skipChars(nChars: Long, finfo: FormatInfo): Boolean = {
+    // threadCheck()
+    val startingBitPos = bitPos0b
+
+    val aligned = align(finfo.encodingMandatoryAlignmentInBits, finfo)
+    if (!aligned) {
+      false
+    } else {
+      var remainingCharsToSkip = nChars
+      var keepGoing = true
+
+      while (keepGoing && remainingCharsToSkip > 0) {
+        val charsToSkip = Math.min(remainingCharsToSkip, skipCharBuffer.capacity)
+        skipCharBuffer.position(0)
+        skipCharBuffer.limit(charsToSkip.toInt)
+
+        val numDecoded = finfo.decoder.decode(this, finfo, skipCharBuffer)
+        remainingCharsToSkip -= numDecoded
+
+        if (numDecoded == 0) {
+          keepGoing = false
+        }
+      }
+
+      val skippedAllNChars = remainingCharsToSkip == 0
+
+      if (!skippedAllNChars) {
+        // failed to skip all the necessary chars, reset the bit position back to
+        // where we started
+        setBitPos0b(startingBitPos)
+      }
+
+      skippedAllNChars
+    }
+  }
+
+  def lookingAt(matcher: java.util.regex.Matcher, finfo: FormatInfo): Boolean = {
```

Comment:
Did this code change much? Fundamental algorithm is still gradually enlarging the buffer until big enough to hold a match or not. 

I can''t see deltas because the diff tool isn't relating this to the prior existing code.

Reply:
Not really. It's simplified some because it used to have logic about looping the fillCharBuffer since decode would stop at the first error, but now we just always replace things. There also was a lot of logic for determine lengths and that completely goes away since the decoder now keeps track of the bit position where each decoded char ends. Also I think some minor performance gains like allocation of a tuple no longer happens. But otherwise, the logic is virtually the same as before. It decodes some amount, tries a match, if that doesn't work, decode more, and repeat.